File: androidApp/src/main/java/dev/haroldjose/familysharedlist/android/presentationLayer/pages/quickInsertList/viewmodesls/QuickInsertListViewModel.kt
package dev.haroldjose.familysharedlist.android.presentationLayer.pages.quickInsertList.viewmodesls

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import dev.haroldjose.familysharedlist.Logger
import dev.haroldjose.familysharedlist.domainLayer.models.FamilyListModel
import dev.haroldjose.familysharedlist.domainLayer.usecases.familyList.CreateFamilyListUseCase

class QuickInsertListViewModel(
    private val createFamilyListUseCase: CreateFamilyListUseCase
): ViewModel(), IQuickInsertListViewModel {
    override var viewState: QuickInsertListViewState = QuickInsertListViewState.Initial
    override var text:String by mutableStateOf("")
    override var goToFamilyListPage: () -> Unit = {}
    override suspend fun quickInsertItem() {

        if (text.isEmpty()) return

        val listOfItem = text.lines().mapNotNull { item ->

            //extrai os numeros
            val quantityStr = item.filter { it.isDigit() }.toIntOrNull()
            //extrai o nome
            val name = item.filter { !it.isDigit() }

            if (name.isEmpty())
                return@mapNotNull null

            return@mapNotNull FamilyListModel(
                name = name.ifEmpty { item },
                isPrioritized = true,
                quantity = quantityStr ?: 1
            )
        }
        text = ""
        viewState = QuickInsertListViewState.Loading
        try {
            createFamilyListUseCase.execute(items = listOfItem)
            goToFamilyListPage()
            QuickInsertListViewState.Success
        } catch (e: Throwable) {
            showError(e)
            return
        }
    }

    private fun showError(e: Throwable) {
        e.message?.let { Logger.d("showError", it) }
        viewState = QuickInsertListViewState.Error(
            message = e.message ?: "Erro inesperado",
            retryAction = { viewState = QuickInsertListViewState.Initial }
        )
    }
}

File: iosApp/iosApp/presentationLayer/pages/quickInsertList/viewModels/QuickInsertListViewModel.swift
import Foundation
import SwiftUI
import shared
import KMPNativeCoroutinesAsync

class QuickInsertListViewModel: QuickInsertListViewModelProtocol {

    @Published var viewState: QuickInsertListViewState = .initial
    @Published var text = "" {
        didSet {
            if text.isEmpty {
                textFieldTitle = "Informe os itens sendo cada linha um novo item."
            } else {
                let count = text.components(separatedBy: "\n").count
                let itemText = count == 1 ? "item" : "itens"
                textFieldTitle = "\(count) \(itemText)"
            }
        }
    }
    @Published var textFieldTitle = "Informe os itens sendo cada linha um novo item."
    var goToFamilyListPage: () -> Void = {}

    private let createFamilyListUseCase: CreateFamilyListUseCase
    private let crashlytics: IFirebaseCrashlytics

    init(
        createFamilyListUseCase: CreateFamilyListUseCase,
        crashlytics: IFirebaseCrashlytics
    ) {
        self.createFamilyListUseCase = createFamilyListUseCase
        self.crashlytics = crashlytics
    }

    @MainActor
    func quickInsertItem() async {
        guard !text.isEmpty else { return }

        let listOfItem: [FamilyListModel] = text.components(separatedBy: .newlines).compactMap({ item in
            let quantityStr = item.filter { $0.isNumber }
            let name = item.filter { !$0.isNumber }

            if name.isEmpty {
                return nil
            }

            return FamilyListModel(
                name: name.isEmpty ? item : name,
                isPrioritized: true,
                quantity: Int32(quantityStr) ?? 1
            )
        })

        text = ""
        viewState = .loading

        do {
            _ = try await asyncFunction(for: createFamilyListUseCase.execute(items: listOfItem))
            viewState = .success
            goToFamilyListPage()
        } catch {
            showError(e: error)
            return
        }
    }

    func showError(e: Error) {
        crashlytics.record(error: e)
        viewState = .error(message: e.localizedDescription, retryAction: {
            self.viewState = .initial
        })
    }
}
